crates/build/src/lib.rs

```diff
@@ -6,25 +6,49 @@ mod manifest;
 
 use anyhow::{anyhow, bail, Context, Result};
 use spin_loader::local::parent_dir;
-use std::path::{Path, PathBuf};
+use std::{
+    collections::HashSet,
+    path::{Path, PathBuf},
+};
 use subprocess::{Exec, Redirection};
 
 use crate::manifest::{BuildAppInfoAnyVersion, RawComponentManifest};
 
 /// If present, run the build command of each component.
-pub async fn build(manifest_file: &Path) -> Result<()> {
+pub async fn build(manifest_file: &Path, component_ids: &[String]) -> Result<()> {
     let manifest_text = tokio::fs::read_to_string(manifest_file)
         .await
         .with_context(|| format!("Cannot read manifest file from {}", manifest_file.display()))?;
     let app = toml::from_str(&manifest_text).map(BuildAppInfoAnyVersion::into_v1)?;
     let app_dir = parent_dir(manifest_file)?;
 
-    if app.components.iter().all(|c| c.build.is_none()) {
-        println!("No build command found!");
+    let components_to_build = if component_ids.is_empty() {
+        app.components
+    } else {
+        let all_ids: HashSet<_> = app.components.iter().map(|c| &c.id).collect();
+        let unknown_component_ids: Vec<_> = component_ids
+            .iter()
+            .filter(|id| !all_ids.contains(id))
+            .map(|s| s.as_str())
+            .collect();
+
+        if !unknown_component_ids.is_empty() {
+            bail!("Unknown component(s) {}", unknown_component_ids.join(", "));
+        }
```

Comment:
nit: Might as well avoid the unnecessary allocation in the happy path?

```suggestion
        let unknown_component_ids = component_ids
            .iter()
            .filter(|id| !all_ids.contains(id))
            .peekable();

        if unknown_component_ids.peek().is_some() {
            bail!("Unknown component(s) {}", unknown_component_ids.collect::<Vec<_>>().join(", "));
        }
```

Reply:
TIL: `peekable()` - thank you!

"Avoid allocating an empty vector" doesn't feel like a huge priority when the happy path goes on to run an external process (a compiler, no less).  And I have to admit I find `is_empty` more obvious than `peek().is_some()`, though maybe the latter is idiomatic to Rust devs?

But I can change it (to this or @lann's suggestion) if other folks prefer.  We could also use @lann's mention of `itertools::join` to simplify the sad path in this one.

Reply:
No need to block on my nits. I'll let you decide how to proceed from here. 😀